numpyro/contrib/control_flow/scan.py

```diff
@@ -141,35 +149,54 @@ def body_fn(wrapped_carry, x, prefix=None):
                 seeded_fn = handlers.substitute(seeded_fn, substitute_fn=subs_fn)
 
         if init:
-            with handlers.scope(prefix="_init"):
-                new_carry, y = seeded_fn(carry, x)
+            # handler the name to match the pattern of sakkar_bilmes product
+            with handlers.scope(prefix='P' * (unroll_steps - i), divider='_'):
+                new_carry, y = config_enumerate(seeded_fn)(carry, x)
                 trace = {}
         else:
-            with handlers.block(), packed_trace() as trace, promote_shapes(), enum(), markov():
```

Comment:
The usage of `enum()`, `markov()` here still gives correct log density but does not recycle the dimensions at this step. To resolve this, I have moved `block` right before the generator `markov(unroll_steps + 1, history)` and remove `markov` handler here.